src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -189,35 +189,51 @@ export class LocalKernelFinder implements ILocalKernelFinder {
         let filteredInterpreters = [...interpreters];
 
         // Then go through all of the kernels and generate their metadata
-        const kernelMetadata = kernelSpecs.map((k) => {
-            // Find the interpreter that matches. If we find one, we want to use
-            // this to start the kernel.
-            const matchingInterpreters = this.findMatchingInterpreters(k, interpreters);
-            if (matchingInterpreters && matchingInterpreters.length) {
-                const result: PythonKernelConnectionMetadata = {
-                    kind: 'startUsingPythonInterpreter',
-                    kernelSpec: k,
-                    interpreter: matchingInterpreters[0],
-                    id: getKernelId(k, matchingInterpreters[0])
-                };
-
-                // If interpreters were found, remove them from the interpreter list we'll eventually
-                // return as interpreter only items
-                filteredInterpreters = filteredInterpreters.filter((i) => !matchingInterpreters.includes(i));
+        const kernelMetadata = await Promise.all(
+            kernelSpecs.map(async (k) => {
+                // Find the interpreter that matches. If we find one, we want to use
+                // this to start the kernel.
+                const matchingInterpreters = this.findMatchingInterpreters(k, interpreters);
```

Comment:
findMatchingInterpreters is supposed to look at the kernel metadata. How come it didn't find it?

Reply:
Meaning your fix should be redundant.

Reply:
Because the interpreter is a virtual environment in a workspace that not visible in another workspace.

E.g. Worksapce A - Virtual Env - PythonABC
Open sub directory in Worksapce A or anotehr worksapce, now the Virtual Env Python ABC is not listed as a python environment. But this is listed as a valid kernel because we have regsitered this kernel.

So, when you go into anotehr worksapce folder and open an existing notebook or or a new notebook & select an existing kernel, it will not work.

Reply:
i.e. workspace virtual envs don't appear in all worksapces in the interpreter list, but kernels pecs do (as they are registered gloablly)

Reply:
So getInterpreterDetails will work even though it's not an allowed interpreter?

Reply:
If that's true, it seems like we should be calling that in findMatchingInterpreters?

Reply:
Actually no where you have it seems better to me. Maybe a comment saying the reasons why this can happen though would be nice. 

The call to findResourceInterpreters doesn't find interpreters that are global.

Reply:
> So getInterpreterDetails will work even though it's not an allowed interpreter?

Its not correct to say `allowed interrpeter`, its basically an interpreter that wasn't discovered.

Reply:
Will update comments.

Reply:
Done